### PR TITLE
Predict class using SVM

### PR DESCRIPTION
**Pull request by murphymadeleine21**

### CODE REVIEW

**Comment by murphymadeleine21** [comment]
I played around with svm.SVC today and worked through the documentation and various examples to come up with this code.. Originally I had a very simple model working with kernel="linear". However, I noticed that if I changed the kernel to be poly or RBF, the accuracy improved greatly (linear ~ 0.5 accuracy). Then I played around with GridSearchSV as it was recommended in the documentation to find the best C and gamma values. The accuracy using RBF has gotten as high as .80 but usually hovers between .65-.8 from what I saw. I wasn't sure if we were just trying to do a simple linear model, or if it was okay to explore and see what sort of different outputs we could get, but this is essentially my first draft of predicting using the SVC function. 
Additionally, I noticed that when I ran kernel="linear", all predicted values for both CP and NV were "1". But when you run with poly or rbf, you get a mix of 1s and 0s, which is obviously more ideal.

**Comment by aarmey** [comment]
You can definitely use RBF. Not surprising that it's better.

File: syserol/model.py [hunk 1]
```diff
@@ -126,3 +126,52 @@ def cross_validation():
...
+    
+    # Controller/Progressor classification
+    # Split data into training and testing data 
+    X_train, X_test, y_train, y_test = train_test_split(subjects_matrix, cp)
+
+    # Kernel = RBF
+    # Search for best RBF parameters for estimation
+    C_range = np.logspace(-2, 10, 13)
+    gamma_range = np.logspace(-9, 3, 13)
+    param_grid = dict(gamma=gamma_range, C=C_range)
```
Comment (aarmey) [inline syserol/model.py]: Don't use grid search. Just use the parameter defaults.

File: syserol/model.py [hunk 2]
```diff
@@ -126,3 +126,52 @@ def cross_validation():
...
+    # Viremic/Nonviremic classification
+    # Split data into training and testing data 
+    X_train, X_test, y_train, y_test = train_test_split(subjects_matrix, nv)
+
+    # Kernel = RBF
+    # Search for best RBF parameters for estimation
+    C_range = np.logspace(-2, 10, 13)
+    gamma_range = np.logspace(-9, 3, 13)
+    param_grid = dict(gamma=gamma_range, C=C_range)
+    grid = GridSearchCV(SVC(), param_grid=param_grid, cv=10) # 10 fold cross validation to find parameters
```
Comment (aarmey) [inline syserol/model.py]: Same here.

File: syserol/model.py [hunk 3]
```diff
@@ -126,3 +126,52 @@ def cross_validation():
...
+    nv = np.array(classes["class.nv"])
+    
+    # Run Decomposition (5 components)
+    tensorFac, matrixFac, R2XX = perform_CMTF(cube, glyCube, 5)
+    # Subject Matrix acts for model training data
+    subjects_matrix = tensorFac[1][0]
+    
+    # Controller/Progressor classification
+    # Split data into training and testing data 
+    X_train, X_test, y_train, y_test = train_test_split(subjects_matrix, cp)
```
Comment (aarmey) [inline syserol/model.py]: You only have one CV fold? Should use `cross_val_predict` here.

**Review by aarmey — CHANGES_REQUESTED** [review]

File: syserol/dataImport.py [hunk 4]
```diff
@@ -92,12 +92,9 @@ def createCube():
...
     dfGlycan = dfGlycan.pivot(index="subject", columns="variable", values="value")
     func, _ = importFunction()
-    classes = load_file("meta-subjects")
-    classes = classes.drop(["class.etuv"], axis = 1)
-    classes = classes.replace(to_replace=["controller", "progressor", "viremic", "nonviremic"], value = [1, 0, 1, 0])
-    data_frames = [dfGlycan, func, classes]
+    data_frames = [dfGlycan, func]
     df_merged = reduce(lambda left, right: pd.merge(left, right, on=["subject"], how="outer"), data_frames)
-    glyCube = np.full([len(subjects), len(glycan)+8], np.nan)
+    glyCube = np.full([len(subjects), len(glycan)+6], np.nan)
```
Comment (aarmey) [inline syserol/dataImport.py]: Put this change on a separate branch. We should check that it doesn't affect any other results.

File: syserol/model.py (hunk 3 above)
Comment (murphymadeleine21) [reply syserol/model.py]: Okay let me know if the fixes below are how you wanted it!

**Comment by aarmey** [comment]
This looks correct, but we'll need to evaluate the other PR first.

File: syserol/model.py [hunk 5]
```diff
@@ -52,3 +56,42 @@ def cross_validation():
         matrix[test_index, 6:13] = pred_matrix[test_index, len(glycan) : len(glycan) + 6]  # store predicted values
 
     return matrix
+
+def class_predictions():
```
Comment (aarmey) [inline syserol/model.py]: Just take `subjects_matrix` as an argument to this function. A function should do one, focused thing.

Comment (aarmey) [reply syserol/model.py]: This function also needs a more descriptive name.

File: syserol/model.py [hunk 6]
```diff
@@ -52,3 +56,42 @@ def cross_validation():
...
+    # Controller/Progressor classification
+    X = subjects_matrix
+    Y = cp
+    # Kernel = RBF
+    # Run SVM classifier model
+    clf = SVC(kernel="rbf")
+    y_pred1 = cross_val_predict(clf, X, Y, cv=10)
+    model = clf.fit(X, Y) # fit model
+    cp_accuracy = accuracy_score(Y, y_pred1)
+    print("CP Prediction Accuracy:", cp_accuracy)
```
Comment (aarmey) [inline syserol/model.py]: Don't print within here. Output should be in the figure.

File: syserol/model.py [hunk 7]
```diff
@@ -52,3 +56,42 @@ def cross_validation():
...
+    
+    
+    # Viremic/Nonviremic classification 
+    X = subjects_matrix
```
Comment (aarmey) [inline syserol/model.py]: X doesn't change from above. Do you need this line?

File: syserol/model.py [hunk 8]
```diff
@@ -52,3 +56,42 @@ def cross_validation():
...
+    subjects_matrix = tensorFac[1][0]
+    
+    # Controller/Progressor classification
+    X = subjects_matrix
+    Y = cp
+    # Kernel = RBF
+    # Run SVM classifier model
+    clf = SVC(kernel="rbf")
+    y_pred1 = cross_val_predict(clf, X, Y, cv=10)
+    model = clf.fit(X, Y) # fit model
```
Comment (aarmey) [inline syserol/model.py]: Why do you run fit? model isn't used.

File: syserol/model.py [hunk 9]
```diff
@@ -52,3 +56,42 @@ def cross_validation():
...
+    
+    
+    # Viremic/Nonviremic classification 
+    X = subjects_matrix
+    Y = nv
+    # Kernel = RBF
+    # Run SVM classifier model
+    clf = SVC(kernel="rbf")
+    y_pred2 = cross_val_predict(clf, X, Y, cv=10)
+    model = clf.fit(X, Y) # fit model
```
Comment (aarmey) [inline syserol/model.py]: model is never used.

**Review by aarmey — CHANGES_REQUESTED** [review]

**Review by aarmey — APPROVED** [review]
